meta_client.go

```diff
@@ -0,0 +1,386 @@
+}
+
+func NewMetaClient(address HostAddress, timeout time.Duration) *metaClient {
+	return &metaClient{
+		address:             address,
+		timeout:             timeout,
+		latestSchemaVersion: -1,
+		meta:                nil,
+	}
+}
+
+// open the metaClient socket connection
+func (client *metaClient) Open() error {
+	// Process domain to IP
+	var addresses []HostAddress
+	addresses = append(addresses, client.address)
+	convAddress, err := DomainToIP(addresses)
+	if err != nil {
+		return fmt.Errorf("failed to find IP, error: %s ", err.Error())
+	}
+
+	// Check input
+	if len(convAddress) == 0 {
+		return fmt.Errorf("failed to initialize connection pool: illegal address input %s:%d", client.address.Host, client.address.Port)
+	}
+
+	ip := convAddress[0].Host
+	port := convAddress[0].Port
+	return client.doOpen(ip, port)
+}
+
+// do open operation for metaClient
+func (client *metaClient) doOpen(ip string, port int) error {
+	newAdd := net.JoinHostPort(ip, strconv.Itoa(port))
+
+	sock, err := thrift.NewSocket(thrift.SocketAddr(newAdd), thrift.SocketTimeout(client.timeout))
+	if err != nil {
+		return fmt.Errorf("failed to create a net.Conn-backed Transport,: %s", err.Error())
+	}
+
+	bufferSize := 128 << 10
+	bufferedTranFactory := thrift.NewBufferedTransportFactory(bufferSize)
+	transport := thrift.NewHeaderTransport(bufferedTranFactory.GetTransport(sock))
+	pf := thrift.NewHeaderProtocolFactory()
+
+	client.meta = meta.NewMetaServiceClientFactory(transport, pf)
+	if err = client.meta.Open(); err != nil {
+		return fmt.Errorf("failed to open transport, error: %s", err.Error())
+	}
+
+	if !client.meta.IsOpen() {
+		return fmt.Errorf("transport is off")
+	}
+	return client.verifyClientVersion()
+}
+
+// verify the compatibility of client and server
+func (client *metaClient) verifyClientVersion() error {
+	req := meta.NewVerifyClientVersionReq()
+	resp, err := client.meta.VerifyClientVersion(req)
+	if err != nil {
+		client.Close()
+		return fmt.Errorf("failed to verify client version: %s", err.Error())
+	}
+	if resp.GetCode() != nebula.ErrorCode_SUCCEEDED {
+		client.Close()
+		return fmt.Errorf("incompatible version between client and server: %s", string(resp.GetErrorMsg()))
+	}
+	return nil
+}
+
+// get all space names
+func (client *metaClient) GetSpaces() ([]string, error) {
+	req := meta.NewListSpacesReq()
+
+	resp, err := client.meta.ListSpaces(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// fresh metaClient with real leader and re-execute the request
+	if resp.GetCode() == nebula.ErrorCode_E_LEADER_CHANGED {
+		client.freshClient(resp.GetLeader())
+		resp, err = client.meta.ListSpaces(req)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if resp.GetCode() == nebula.ErrorCode_SUCCEEDED {
+		var spaces []string
+		for _, name := range resp.GetSpaces() {
+			spaces = append(spaces, string(name.GetName()))
+		}
+		return spaces, nil
+	} else {
+		return nil, fmt.Errorf("GetSpaces failed, code:%s", resp.GetCode())
+	}
+}
```

Comment:
How about return error first to reduce netsting else statement?

Reply:
> How about return error first to reduce netsting else statement?

good idea, will change it.